format/spec.md

```diff
@@ -842,7 +842,8 @@ The rows in the delete file must be sorted by `file_path` then `pos` to optimize
 
 Equality delete files identify deleted rows in a collection of data files by one or more column values, and may optionally contain additional columns of the deleted row.
 
-Equality delete files store any subset of a table's columns and use the table's field ids. The _delete columns_ are the columns of the delete file used to match data rows. Delete columns are identified by id in the delete file [metadata column `equality_ids`](#manifests). Float and double columns cannot be used as delete columns in equality delete files.
+Equality delete files store any subset of a table's columns and use the table's field ids. The _delete columns_ are the columns of the delete file used to match data rows. Delete columns are identified by id in the delete file [metadata column `equality_ids`](#manifests). The column restrictions for columns used in equality delete files are the same as those for [identifier fields](#identifier-field-ids) with the exception that optional columns and columns nested under optional structs are allowed (if a 
+parent struct column is null it implies the leaf column is null).
```

Comment:
I think there is one missing part: how null values treated in equality ids? Are they treat as equal or unequal? Identity ids don't allow null values.

Reply:
What would be the meaning of a null in the list of equality IDs? If I'm not mistaken equality IDs are used for identifying which columns are used for the equality checks, but 'null column' doesn't make sense for me.

If you mean null values in the equality dele files, that's a good question. I believe in SQL world NULL doesn't equal to any value, even NULL doesn't equal to NULL, so I wonder what would be the desired semantics when we find a NULL value in the equality delete file. Should we apply "IS NULL" on that particular column? But then it won't be an equality check as the name "equality delete" would suggest but an IS NULL check. I'd simply not allow NULL values in the delete files TBH.

Reply:
The definition of null values present is already explained in the paragraph starting on line 850, please let me know if you think this this is insufficient, or if there is some word smithing needed (I'm open to suggestions).

Reply:
Sorry I missed the paragraph later. I think the statement is clear enough that a `NULL` values means `is not NULL`. I think it's reasonable to have that definition in equality check.

Reply:
just to clarify I think it means NULL means `is null` (i'm guess `is not null` is just a typo)

Reply:
> just to clarify I think it means NULL means `is null` (i'm guess `is not null` is just a typo)

Yes, it's a typo.

Reply:
I have one implementation related objection against treating `NULL`s in the delete files as `IS NULL`s:
This might be different in each query engine but in general I think it's in planning time when you create the predicates for the query, but the information that a row in the delete file contains `NULL` will be known only during the execution phase when you read the actual rows. But at that point you won't be exchanging an equality predicate to an `IS NULL` predicate, right?
Do you know if query engines already are able to get around this problem and treat `NULL` values in the delete files as `IS NULL`?

Reply:
I think this is guaranteed if all query engine uses `DeletionFilter` provide in iceberg.

Reply:
I don't think we can take it guaranteed that all the engines use `DeleteFilter`. For instance Impala doesn't.

Reply:
Yes, that's a problem of different  language implementation.

Reply:
@gaborkaszab I think the concern over nulls should be brought to a different forum as it is already part of the spec unambiguously.  

I think the implications for the predicates if you are generating them as part of a query plan is that they for nullable columns they need to be written as `(delete_file_column is not null && data_file_column = delete_file_column) || (delete file column is null && data file column is null)` for matching rows that should be deleted.

Reply:
You're right, this predicate would do the trick for handling null columns too. Thanks!

Reply:
> with the exception that optional columns and columns nested under optional structs are allowed (if a parent struct column is null it implies the leaf column is null).

Is that true? If does not make sense to me. I would not expect a null value in an identifier column (as also stated above).

Reply:
@Fokko sorry I missed the question.  this might be poor wording on my part but:
1.  For identifier columns, nulls are not allowed (this is already part of the spec)
2. For equality deletes null columns are allowed (either top-level or via nesting).  Nullability of equality delete columns are already explicitly allowed in the specification.

Reply:
Thanks for the clarification, this makes sense to me 👍